asdf/asdf.py

```diff
@@ -908,51 +909,57 @@ def _open_impl(
         **kwargs,
     ):
         """Attempt to open file-like object as either AsdfFile or AsdfInFits"""
+        close_on_fail = isinstance(fd, (str, pathlib.Path))
         generic_file = generic_io.get_file(fd, mode=mode, uri=uri)
-        file_type = util.get_file_type(generic_file)
-
-        if file_type == util.FileType.FITS:
-            try:
-                # TODO: this feels a bit circular, try to clean up. Also
-                # this introduces another dependency on astropy which may
-                # not be desireable.
-                from . import fits_embed
+        try:
```

Comment:
I am not a big fan of putting yet another try/except in this method. I find it difficult to follow the logic of why we are catching errors raised within the function itself.

I suggest refactoring slightly, where we split part of your version of `_open_mpl` into a new method,say `_open_base` (think of a better name if you can).

I would keep `_open_mpl` with the same interface, but use it as a wrapper with your try/except around `_open_base` i.e:
```python
close_on_fail = isinstance(fd, (str, pathlib.Path))
generic_file = generic_io.get_file(fd, mode=mode, uri=uri)

try:
    return cls._open_base(..., generic_file,....)
except Exception as e:
    if close_on_fail:
        generic_file.close()
    raise e
```
Then `_open_base` would essentially be the "original" `_open_mpl`, these parts: https://github.com/asdf-format/asdf/blob/2b4afb65e932a3fcd39056efa221857ee798ae96/asdf/asdf.py#L914-L955  But with interface modified so that it accepts `generic_file` directly rather than calling `get_file` on something.

Doing this will make it much more clear that we are specifically trying to close the file after an error is raised during a file open. This can be documented in the document strings for the two functions.

Reply:
Note that in general, I find tracing the logic of opening the asdf file through each of the open functions involved within `asdf.py` quite tedious. I'd like to see those functions refactored into more, but smaller logical units.

Reply:
Thanks for the descriptions. Refactoring the open functions sounds like a great idea.

Currently open (for an asdf file) calls:
- open_asdf
- _open_impl
- _open_asdf
All 3 of these have overlapping arguments with defaults. Cleaning this up might be best handled by pulling some of those arguments into asdf_config (as mentioned in AL-652: Move remaining asdf.open and AsdfFile configuration options to AsdfConfig).

Reply:
This clean up should probably go into a separate PR. It might be a good intermediate project. We should look at the 3.0 branch to see if there are any changes that we can pull back to 2.x

Reply:
To clarify the part of the refactor I initially proposed here should stay in this PR. I think the discussion here warrants further discussion. Lets move this to issue #1223.